### PR TITLE
feat: Switch from Yoga v2 to Apollo Server v4

### LINKED ISSUE #8728: Switch from Yoga v2 to Apollo V4
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation

Yoga V4 is currently not suitable for prod. I encountered issues on upload feature, and types issues in a monorepo.

### Feature / Enhancement Description

Use the last version of Apollo Server. It could also help in the futur to get some extra features for developers using apollo client.

### Example Use Case

More details on the linked PR

### Alternatives / Workarounds

None

### 3rd Party References

https://www.apollographql.com/

### PR DESCRIPTION
**Pull request by Moumouls**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

## Issue

Closes: #8728

## Approach

- Install graphql upload to support upload
- Install GraphIQL from Yoga team since GraphIQL is standard, and the yoga team graphiql is fine
- Take advantage of X-Parse-Application-Id as CSRF protection [https://www.apollographql.com/docs/router/configuration/csrf/](https://www.apollographql.com/docs/router/configuration/csrf/)
- apollo/server is more focused on NodeJS and have dedicated expressHandler, i hope it will be more stable in complex envs.
- Now we have end to end stream system from client to bucket/storage system through GQL Upload API
- Disable the cache control since it's not used at the moment and can slightly help for performance
- Remove all it_only_node < 17, now upload works on major NodeJS version without issues
- Added mime type package to auto add extension to filename without extension since it's the current behavior, without this we will have a breaking change

Performance tests:
- Checked performance on ParseServerGraphQL test suite and complete test suite in my company with a lot of graphql requests, performance are close and quite the same.

## Tasks

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, code comments)
- [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [x] Add new Parse Error codes to Parse JS SDK

### CODE REVIEW

**Comment by mtrezza** [comment]
I notice that this also requires Node 16, so we'll wait with merging info alpha until Nov.

**Review by mtrezza — COMMENTED** [review]
Apart from the requirement of Node >= 16, does this have any breaking changes? It's fine if it does, since we can only release it with Parse Server 7 (2024) anyway.

**Comment by Moumouls** [comment]
@mtrezza no breaking changen may be the error message of graphql upload, but developers should not rely on message.

Also i added mime type detection, since it's currently supported.

Yes @mtrezza it's not a problem to keep this PR for 2024 :)

File: package.json [hunk 1]
```diff
@@ -48,6 +48,7 @@
     "lodash": "4.17.21",
     "lru-cache": "9.1.1",
     "mime": "3.0.0",
+    "mime-types": "2.1.35",
```
Comment (mtrezza) [inline package.json]: We already have [mime](https://www.npmjs.com/package/mime), it seems that it's similar to [mime-types](https://www.npmjs.com/package/mime-types). I think if possible we should decide for one of the two if they provide the same functionality.

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by mtrezza** [comment]
Could you suggest an entry for the changelog?

The current title would only be a refactor, but from the PR comment it seems to include many more improvements, so it would be good to add them to the changelog. You don't have to write that into the PR title, I'll add the long version in the commit message and we keep the PR title short.

**Comment by mtrezza** [comment]
@Moumouls While this PR is still "recent" and to prepare it for merging in Nov, do you think you could address the mime dependency issue? Otherwise I'm afraid this PR will go stale and not get merged for a long time if no-one picks it up.

**Comment by Moumouls** [comment]
Hi @mtrezza, we should definitely merge this one. It works really well.

What is the mime dep issue. How is it linked to this Pr ?

Just tell me when I need to refresh the branch, and I'll be happy to fix last details to get it merged !

**Comment by mtrezza** [comment]
See https://github.com/parse-community/parse-server/pull/8727#discussion_r1313639096.

**Comment by Moumouls** [comment]
Nice catch @mtrezza , I'll take a look

**Comment by Moumouls** [comment]
Hi @mtrezza i replaced last yoga dep with the CDN version of Apollo Sandbox: https://www.apollographql.com/docs/graphos/explorer/sandbox 

Also fixed your feedback to use current mime dep

My team use this branch in production since now 2 months everything works well.

File: src/GraphQL/ParseGraphQLServer.js [hunk 2]
```diff
@@ -79,35 +93,55 @@ class ParseGraphQLServer {
...
+                headers: {
+                    ...options.headers,
+                    'X-Parse-Application-Id': "${JSON.stringify(this.parseServer.config.appId)}",
+                    'X-Parse-Master-Key': "${JSON.stringify(this.parseServer.config.masterKey)}",
+                },
+              })
+            },
+           });
+           // advanced options: https://www.apollographql.com/docs/studio/explorer/sandbox#embedding-sandbox
+          </script>`
```
Comment (mtrezza) [inline src/GraphQL/ParseGraphQLServer.js:144]: Is this safe?

File: src/GraphQL/loaders/filesMutations.js [hunk 3]
```diff
@@ -1,33 +1,63 @@
...
+import { getExtension } from 'mime';
 import { mutationWithClientMutationId } from 'graphql-relay';
 import Parse from 'parse/node';
 import * as defaultGraphQLTypes from './defaultGraphQLTypes';
 import logger from '../../logger';
 
+// Handle graphql file upload and proxy the file upload
+// to the graphql server url specified in the config
+// We do not call directly createFile from the Parse Server
+// to leverage the standard file upload mechanism
```
Comment (mtrezza) [inline src/GraphQL/loaders/filesMutations.js]:
```suggestion
// Handle GraphQL file upload and proxy file upload to GraphQL server url specified in config;
// `createFile` is not directly called by Parse Server to leverage standard file upload mechanism
```
```suggestion
// Handle graphql file upload and proxy the file upload
// to the graphql server url specified in the config
// We do not call directly createFile from the Parse Server
// to leverage the standard file upload mechanism
```

File: src/GraphQL/ParseGraphQLServer.js [hunk 4]
```diff
@@ -57,8 +58,21 @@ class ParseGraphQLServer {
...
-    const options = await this._getGraphQLOptions();
-    this._server = createServer(options);
+    const { schema, context } = await this._getGraphQLOptions();
+    const apollo = new ApolloServer({
+      csrfPrevention: {
+        // See https://www.apollographql.com/docs/router/configuration/csrf/
+        // needed since we use graphql upload
+        requestHeaders: ['X-Parse-Application-Id'],
+      },
+      introspection: true,
```
Comment (kekami) [inline src/GraphQL/ParseGraphQLServer.js:68]: Introspection should probably be disabled by default for security reasons. Perhaps a config variable.

File: src/GraphQL/ParseGraphQLServer.js (hunk 2 above)
Comment (Moumouls) [reply src/GraphQL/ParseGraphQLServer.js:144]: it's safe @mtrezza , also the playground should be only used in dev, never in production, it's mentionned in the docs

File: src/GraphQL/ParseGraphQLServer.js (hunk 4 above)
Comment (Moumouls) [reply src/GraphQL/ParseGraphQLServer.js:68]: Hi @kekami , you are right but it needs a dedicated PR, i created the FR: https://github.com/parse-community/parse-server/issues/8800

**Comment by mtrezza** [comment]
@Moumouls We have begun merging the breaking change PRs for Parse Server 7. Could you please resolve the conflict in this PR so we can merge it?

**Comment by mtrezza** [comment]
Closing via https://github.com/parse-community/parse-server/pull/8959

**Pull request closed without merging**